chromium_src/chrome/browser/component_updater/widevine_cdm_component_installer.cc

```diff
@@ -3,39 +3,307 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "chrome/browser/component_updater/widevine_cdm_component_installer.h"
+
+#if BUILDFLAG(WIDEVINE_ARM64_DLL_FIX)
+#define WidevineCdmComponentInstallerPolicy \
+  WidevineCdmComponentInstallerPolicy_ChromiumImpl
+#endif
+
 #define RegisterWidevineCdmComponent RegisterWidevineCdmComponent_ChromiumImpl
 #include "src/chrome/browser/component_updater/widevine_cdm_component_installer.cc"
 #undef RegisterWidevineCdmComponent
 
+#if BUILDFLAG(WIDEVINE_ARM64_DLL_FIX)
+#undef WidevineCdmComponentInstallerPolicy
+#endif
+
+#include "base/check.h"
+#include "base/functional/callback.h"
+#include "brave/browser/widevine/buildflags.h"
 #include "brave/browser/widevine/widevine_utils.h"
-#include "chrome/browser/component_updater/component_updater_utils.h"
 #include "components/component_updater/component_updater_service.h"
-#include "extensions/common/constants.h"
 
-namespace component_updater {
+#if BUILDFLAG(WIDEVINE_ARM64_DLL_FIX)
 
-namespace {
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/logging.h"
+#include "base/synchronization/waitable_event.h"
+#include "base/threading/thread_restrictions.h"
+#include "base/time/time.h"
+#include "base/values.h"
+#include "components/update_client/task_traits.h"
+#include "components/update_client/utils.h"
+#include "net/base/net_errors.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "third_party/zlib/google/zip.h"
+#include "url/gurl.h"
 
-void OnWidevineRegistered() {
-  component_updater::BraveOnDemandUpdate(widevine_extension_id);
-}
+#endif
+
+using update_client::CrxInstaller;
 
-void RegisterAndInstallWidevine(ComponentUpdateService* cus) {
-  // This code is similar to RegisterWidevineCdmComponent_ChromiumImpl
-  // but that ignores the callback, and we handle it so we can force
-  // an on demand update.
+namespace component_updater {
+
+#if !BUILDFLAG(WIDEVINE_ARM64_DLL_FIX)
+
+void RegisterWidevineCdmComponent(ComponentUpdateService* cus,
+                                  base::OnceCallback<void()> callback) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (!IsWidevineOptedIn()) {
+    return;
+  }
   auto installer = base::MakeRefCounted<component_updater::ComponentInstaller>(
       std::make_unique<WidevineCdmComponentInstallerPolicy>());
-  installer->Register(cus, base::BindOnce(&OnWidevineRegistered));
+  installer->Register(cus, std::move(callback));
 }
 
-}  // namespace
+#else  // !BUILDFLAG(WIDEVINE_ARM64_DLL_FIX)
+
+const char kArm64DllUrl[] =
+    "https://dl.google.com/widevine-cdm/4.10.2557.0-win-arm64.zip";
```

Comment:
Do we need to update this url whenever latest one is found?

Reply:
Yes, unfortunately. Some Widevine versions are listed on https://dl.google.com/widevine-cdm/versions.txt. I don't know if new entries there will have `-arm64.zip` files. I've set up an alarm so I get notified when the URL changes. In either case, the versions do not seem to change very often.

Reply:
how the usual widevine download/update process works? Why don't we need to hardcode the version there?

Reply:
1. Browser asks Brave's component update server for the latest version of the Widevine component.
2. Brave's component update server redirects to Google's for licensing reasons.
3. Google's component update server returns the latest version of the Widevine component.
4. Browser downloads and installs into the profile directory.

We don't need to hardcode the version there because the component update server returns it.

The main reason for the workaround implemented in this PR are licensing restrictions. We are not allowed to distribute Widevine ourselves. Without this constraint, we could just ship our own version of the Widevine component from our component update server. But as it is, there is no public endpoint (that I'm aware of, at least) that can be polled for the latest version of `-arm64.zip`. Furthermore, the version of `-arm64.zip` does not coincide with the component version. For example, I have version `4.10.2652.1` of the component with the `2557` `arm64.zip` and it works. But there is no `2652` `arm64.zip`.

Reply:
Can I mark this as resolved @goodov @simonhong?

Reply:
This can also be solved with Griffin: add `base::Feature` with a string parameter. This way we can update this URL without rebuilding. You can set the default parameter value to the current URL.

Reply:
FYI https://github.com/brave/brave-core/pull/18865 is good example for setting default parameter.

Reply:
I switched the implementation to Griffin in a1ea1eee0aff0c5a1455866406995bb0076584de. Wdyt @simonhong @goodov?

Reply:
Looks great. Thanks!